x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -93,29 +102,39 @@
+    void resolve(TransportRequest request, Metadata metadata, AsyncSupplier<List<String>> authorizedIndicesSupplier,
+                 ActionListener<ResolvedIndices> listener) {
         if (request instanceof IndicesAliasesRequest) {
-            ResolvedIndices.Builder resolvedIndicesBuilder = new ResolvedIndices.Builder();
             IndicesAliasesRequest indicesAliasesRequest = (IndicesAliasesRequest) request;
+            GroupedActionListener<ResolvedIndices> groupedListener = new GroupedActionListener<>(listener.map(allResolvedIndices -> {
+                ResolvedIndices.Builder resolvedIndicesBuilder = new ResolvedIndices.Builder();
+                for (ResolvedIndices resolvedIndices : allResolvedIndices) {
+                    resolvedIndicesBuilder.add(resolvedIndices);
+                }
+                return resolvedIndicesBuilder.build();
+            }), indicesAliasesRequest.getAliasActions().size());
             for (IndicesRequest indicesRequest : indicesAliasesRequest.getAliasActions()) {
-                final ResolvedIndices resolved = resolveIndicesAndAliases(indicesRequest, metadata, authorizedIndices);
-                resolvedIndicesBuilder.addLocal(resolved.getLocal());
-                resolvedIndicesBuilder.addRemote(resolved.getRemote());
+                resolveIndicesAndAliases(indicesRequest, metadata, authorizedIndicesSupplier, groupedListener);
+            }
+        } else {
+            // if for some reason we are missing an action... just for safety we'll reject
+            if (request instanceof IndicesRequest == false) {
+                listener.onFailure(new IllegalStateException("Request [" + request + "] is not an Indices request, but should be."));
+            } else {
+                resolveIndicesAndAliases((IndicesRequest) request, metadata, authorizedIndicesSupplier, listener);
             }
-            return resolvedIndicesBuilder.build();
-        }
-
-        // if for some reason we are missing an action... just for safety we'll reject
-        if (request instanceof IndicesRequest == false) {
-            throw new IllegalStateException("Request [" + request + "] is not an Indices request, but should be.");
         }
-        return resolveIndicesAndAliases((IndicesRequest) request, metadata, authorizedIndices);
     }
 
+    ResolvedIndices resolve(TransportRequest request, Metadata metadata, List<String> authorizedIndices) {
+        PlainActionFuture<ResolvedIndices> future = PlainActionFuture.newFuture();
+        // if the supplier is not async, the method is not async, hence get is non-blocking
+        resolve(request, metadata, listener -> listener.onResponse(authorizedIndices), future);
+        return FutureUtils.get(future, 0, TimeUnit.MILLISECONDS);
+    }
 
-    ResolvedIndices resolveIndicesAndAliases(IndicesRequest indicesRequest, Metadata metadata, List<String> authorizedIndices) {
```

Comment:
This got changed a lot in order to support the `AsyncSupplier<List<String>>` instead of `List<String>`.